### ISSUE TITLE
CloudProvider:Unknown with latest ECS optimized AMI

### ISSUE THREAD

**Issue by Mortinke** [google/cadvisor#2096]
We run cAdvisor as daemon on each EC2 instances in all our ECS clusters. Currently, the cAdvisor can't get the informaiton about the CloudProvider, InstanceType and InstanceID. I already tried to provide the EC2 instance ID in the /etc/machine-id file, but this doesn't help:
```
wget -q -O - http://169.254.169.254/latest/meta-data/instance-id > /etc/machine-id
``` 

Unfortunately, I did not find any runtime parameter to specify any of this information.

latest Amazon AMI for eu-central-1 (ami-0b9fee3a2d0596ed1)
```
$ cat /etc/system-release
Amazon Linux AMI release 2018.03
```

```
docker inspect 159e676447d6
[
    {
        "Id": "159e676447d676a8a4c502a98564fbc289697a387f18c83795d5295db7e36639",
        "Created": "2018-10-22T11:24:31.700393406Z",
        "Path": "/usr/bin/cadvisor",
        "Args": [
            "-logtostderr"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 3159,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2018-10-22T11:24:33.766868901Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:0b11cc259ef734ee31253f05af0d5b7a7d4a8d54f763ae6dc65898970681731b",
        "ResolvConfPath": "/var/lib/docker/containers/159e676447d676a8a4c502a98564fbc289697a387f18c83795d5295db7e36639/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/159e676447d676a8a4c502a98564fbc289697a387f18c83795d5295db7e36639/hostname",
        "HostsPath": "/var/lib/docker/containers/159e676447d676a8a4c502a98564fbc289697a387f18c83795d5295db7e36639/hosts",
        "LogPath": "/var/lib/docker/containers/159e676447d676a8a4c502a98564fbc289697a387f18c83795d5295db7e36639/159e676447d676a8a4c502a98564fbc289697a387f18c83795d5295db7e36639-json.log",
        "Name": "/ecs-com-cadvisor-core-ecs-cluster-task-definition-4-com-cadvisor-core-ecs-cluster-container-definition-d89dedaba08cf296dd01",
        "RestartCount": 0,
        "Driver": "devicemapper",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "/:/rootfs:ro",
                "/var/run:/var/run",
                "/sys:/sys:ro",
                "/var/lib/docker:/var/lib/docker:ro",
                "/dev/disk/:/dev/disk:ro",
                "/cgroup:/sys/fs/cgroup:ro"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {
                "8080/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "8080"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "shareable",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 10,
            "Memory": 314572800,
            "NanoCpus": 0,
            "CgroupParent": "/ecs/2dae9e5f-5362-48a4-bf26-43c7690d4fea",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DiskQuota": 0,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 629145600,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": 0,
            "Ulimits": [
                {
                    "Name": "nofile",
                    "Hard": 4096,
                    "Soft": 1024
                }
            ],
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/asound",
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "DeviceId": "19",
                "DeviceName": "docker-259:2-263270-fbecd84bc0d48e435f4409cf61589c7e6df068d30706014c365c2e29dfa70be4",
                "DeviceSize": "10737418240"
            },
            "Name": "devicemapper"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/",
                "Destination": "/rootfs",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rslave"
            },
            {
                "Type": "bind",
                "Source": "/var/run",
                "Destination": "/var/run",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/sys",
                "Destination": "/sys",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/var/lib/docker",
                "Destination": "/var/lib/docker",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rslave"
            },
            {
                "Type": "bind",
                "Source": "/dev/disk",
                "Destination": "/dev/disk",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Type": "bind",
                "Source": "/cgroup",
                "Destination": "/sys/fs/cgroup",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "159e676447d6",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "8080/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "ECS_CONTAINER_METADATA_URI=http://169.254.170.2/v3/a5317861-2141-461b-9feb-26a0532eb8fa",
                "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=/v2/credentials/******************",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "GLIBC_VERSION=2.28-r0"
            ],
            "Cmd": null,
            "ArgsEscaped": true,
            "Image": "google/cadvisor:v0.31.0",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/usr/bin/cadvisor",
                "-logtostderr"
            ],
            "OnBuild": null,
            "Labels": {
                "com.amazonaws.ecs.cluster": "core-ecs-cluster",
                "com.amazonaws.ecs.container-name": "com-cadvisor-core-ecs-cluster-container-definition",
                "com.amazonaws.ecs.task-arn": "arn:aws:ecs:eu-central-1:******************:task/******************",
                "com.amazonaws.ecs.task-definition-family": "com-cadvisor-core-ecs-cluster-task-definition",
                "com.amazonaws.ecs.task-definition-version": "4"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "a43301d3acb4385cdf31649fc6941ee8ff225b113c5912b2c8fa077542104c02",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "8080/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "8080"
                    }
                ]
            },
            "SandboxKey": "/var/run/docker/netns/a43301d3acb4",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "aea6f5269da72b9c2e3e627fed77b47c0cd6a393b5f849567755b5fabd4f9ce6",
            "Gateway": "172.17.0.1",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "172.17.0.2",
            "IPPrefixLen": 16,
            "IPv6Gateway": "",
            "MacAddress": "******************",
            "Networks": {
                "bridge": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "d98eaf2c395ddcde4afd50b238d15a49ef6c6fe61a4bbb2eeef7956f75ba1212",
                    "EndpointID": "aea6f5269da72b9c2e3e627fed77b47c0cd6a393b5f849567755b5fabd4f9ce6",
                    "Gateway": "172.17.0.1",
                    "IPAddress": "172.17.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "******************",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

Could you plese help which configuration changes are necessary to gather the information about CloudProvider, InstanceType and InstanceID?

**Comment by dashpole** [comment]
It shouldn't require configuration; it is just obtained from here: https://github.com/google/cadvisor/blob/master/utils/cloudinfo/aws.go.

**Comment by Mortinke** [comment]
Hi @dashpole,

thanks for your feedback and sorry for the delay. I tried to find out why this code is not executed correctly/not executed at all with `--v=10` but I couldn't find anything. Is there another way to debug/find out why this doesn't work e.g. with another logging configuration?

Thanks in advance!

**Comment by dashpole** [comment]
> Couldn't collect info from any of the files in "/rootfs/etc/machine-id,/var/lib/dbus/machine-id"

Interesting that it is looking in `/rootfs/...` instead of `/etc/...`.  It looks like cAdvisor is detecting [inHostNamespace](https://github.com/google/cadvisor/blob/master/machine/info.go#L97) incorrectly, and adding /rootfs to the beginning of the path

**Comment by dashpole** [comment]
It looks like we assume we are are in the host namespace if we can't find the directory `/rootfs/proc`: https://github.com/google/cadvisor/blob/master/manager/manager.go#L196.  Does that path exist for you?

**Comment by Mortinke** [comment]
@dashpole thanks again for your time. I have a day off today. I will check this tomorrow.

**Comment by Mortinke** [comment]
Yes, the directory exists in the cAdvisor container:
```
/ # ls  /rootfs/proc
1              1344           170            1965           2178           25593          2715           2864           3044           497            crypto         key-users      mtrr           thread-self
10             1353           1702           1966           2199           25634          2720           2865           3045           540            devices        keys           net            timer_list
1008           137            171            1968           2208           25655          27555          2873           3085           6              diskstats      kmsg           pagetypeinfo   tty
1029           14             172            1969           2220           25678          27556          2890           3112           7              dma            kpagecgroup    partitions     uptime
1030           1405           1723           1975           2251           25770          27560          2911           3130           8              driver         kpagecount     sched_debug    version
1071           14532          173            1976           22595          2649           27561          29463          3595           9              execdomains    kpageflags     schedstat      vmallocinfo
109            15             1737           1977           2260           2650           27592          29464          3596           933            filesystems    latency_stats  self           vmstat
11             1562           174            1978           22958          2651           27627          29468          3600           acpi           fs             loadavg        slabinfo       zoneinfo
1138           16             1747           1979           22960          2655           27645          29469          3601           buddyinfo      interrupts     locks          softirqs
12             162            176            1980           22961          2697           280            29515          3654           bus            iomem          mdstat         stat
1293           1653           1781           1985           2373           2699           283            29536          3673           cgroups        ioports        meminfo        swaps
13             16882          18             2              23816          2702           2837           29558          395            cmdline        irq            misc           sys
1306           16922          19             20             25057          2709           2859           3039           4              consoles       kallsyms       modules        sysrq-trigger
1316           16923          1952           2167           25592          2711           2860           3040           401            cpuinfo        kcore          mounts         sysvipc
```

We mount the following directories from the EC2 Instance:

| EC2 path | Container path |
| --------- | ---------------- |
| `/cgroup` | `/sys/fs/cgroup` |
| `/dev/kmsg` | `/dev/kmsg` |
| `/dev/disk/` | `/dev/disk` |
| `/var/lib/docker` | `/var/lib/docker` |
| `/sys` | `/sys` |
| `/var/run` | `/var/run` |
| `/` | `/rootfs` |

**Comment by dashpole** [comment]
> Yes, the directory exists in the cAdvisor container:

Oh, I thought you were running this as a daemon, not a container.  So cAdvisor it is correctly detecting the path: `/rootfs/etc/machine-id`.  So the reason why cAdvisor doesn't have the machine id is because the file `/etc/machine-id` doesn't exist on the host.

We have two options: either create this file on the host, or add to cAdvisor to detect a different file for this  OS.

**Comment by Mortinke** [comment]
Sorry @dashpole, I did not realize that you had answered. What information do you need in the `/etc/machine-id` file? As I wrote in my first comment, I already created the file with `wget -q -O - http://169.254.169.254/latest/meta-data/instance-id > /etc/machine-id` on the host, but nothing changed.

**Comment by dashpole** [comment]
thats strange.  Can you exec into the cAdvisor container and make sure the file is there?  It only tries to get that info on startup, so you need to restart cAdvisor after creating the file.

**Comment by kevingimbel** [comment]
I came across this issue because I was unable to run cadvisor on AWS ECS.

Only after manually creating `/etc/machine-id` cadvisor would not crash, before that it would crash after ~10s, even if run directly on the EC2 instance (with sudo docker run...).

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
